framework/src/modules/token/cc_method.ts

```diff
@@ -22,182 +22,287 @@ import {
 import { EscrowStore } from './stores/escrow';
 import { UserStore } from './stores/user';
 import { InteroperabilityMethod } from './types';
+import { BeforeCCCExecutionEvent } from './events/before_ccc_execution';
+import { RecoverEvent } from './events/recover';
+import { EMPTY_BYTES } from '../interoperability/constants';
+import { BeforeCCMForwardingEvent } from './events/before_ccm_forwarding';
+import { MODULE_NAME_TOKEN } from '../interoperability/cc_methods';
+
+const CHAIN_ID_ALIAS_NATIVE = Buffer.alloc(0); // To be removed
 
 export class TokenInteroperableMethod extends BaseInteroperableMethod {
+	private readonly _tokenMethod: TokenMethod;
+
 	private _interopMethod!: InteroperabilityMethod;
 
-	public constructor(stores: NamedRegistry, events: NamedRegistry, _tokenMethod: TokenMethod) {
+	public constructor(stores: NamedRegistry, events: NamedRegistry, tokenMethod: TokenMethod) {
 		super(stores, events);
+		this._tokenMethod = tokenMethod;
 	}
 
 	public addDependencies(interoperabilityMethod: InteroperabilityMethod) {
 		this._interopMethod = interoperabilityMethod;
 	}
 
-	public async beforeApplyCCM(_ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
-		// const { ccm } = ctx;
-		// const methodContext = ctx.getMethodContext();
-		// if (ccm.fee < BigInt(0)) {
-		// 	throw new Error('Fee must be greater or equal to zero.');
-		// }
-		// const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
-		// const { messageFeeTokenID } = await this._interopMethod.getChannel(
-		// 	methodContext,
-		// 	ccm.sendingChainID,
-		// );
-		// const [feeTokenChainID, feeTokenLocalID] = splitTokenID(messageFeeTokenID);
-		// const userStore = this.stores.get(UserStore);
-		// if (!feeTokenChainID.equals(ownChainID)) {
-		// 	await userStore.addAvailableBalanceWithCreate(
-		// 		methodContext,
-		// 		ctx.trsSender,
-		// 		messageFeeTokenID,
-		// 		ccm.fee,
-		// 	);
-		// 	return;
-		// }
-		// const escrowStore = this.stores.get(EscrowStore);
-		// await escrowStore.deductEscrowAmountWithTerminate(
-		// 	methodContext,
-		// 	this._interopMethod,
-		// 	ccm.sendingChainID,
-		// 	feeTokenLocalID,
-		// 	ccm.fee,
-		// );
-		// const canonicalTokenID = await this._tokenMethod.getCanonicalTokenID(
-		// 	methodContext,
-		// 	messageFeeTokenID,
-		// );
-		// await userStore.addAvailableBalanceWithCreate(
-		// 	methodContext,
-		// 	ctx.trsSender,
-		// 	canonicalTokenID,
-		// 	ccm.fee,
-		// );
+	public async beforeCrossChainCommandExecution(ctx: BeforeApplyCCMsgMethodContext): Promise<void> {
+		const { trsSender, ccm } = ctx;
+		const methodContext = ctx.getMethodContext();
+		const relayerAddress = trsSender;
+		const { id: ownChainID } = await this._interopMethod.getOwnChainAccount(methodContext);
+		const { messageFeeTokenID } = await this._interopMethod.getChannel(
+			methodContext,
+			ccm.sendingChainID,
+		);
+		const { chainID: feeTokenChainID, localID: feeTokenLocalID } = messageFeeTokenID;
```

Comment:
you can use `const [chainID] = splitTokenID(tokenID);` after updating to `getMessageFeeTokenID`